File: .claude/envoy/commands/perplexity.py
"""Perplexity API commands."""

import os
import re

import requests
from .base import BaseCommand
from .xai import XaiSearchCommand


class PerplexityResearchCommand(BaseCommand):
    """Deep research with citations using sonar-deep-research."""

    name = "research"
    description = "Deep research with citations, optional --grok-challenge to validate via X"

    @property
    def timeout_ms(self) -> int:
        # sonar-deep-research is slow, needs longer timeout
        return int(os.environ.get("PERPLEXITY_TIMEOUT_MS", "300000"))

    def add_arguments(self, parser) -> None:
        parser.add_argument("query", help="Research query")
        parser.add_argument("--grok-challenge", action="store_true", help="Challenge findings with Grok X search")

    def execute(self, query: str, grok_challenge: bool = False, **kwargs) -> dict:
        api_key = os.environ.get("PERPLEXITY_API_KEY")
        if not api_key:
            return self.error("auth_error", "PERPLEXITY_API_KEY not set")

        try:
            response, duration_ms = self.timed_execute(
                self._call_api, api_key, query
            )

            content = response["choices"][0]["message"]["content"]
            content = re.sub(r"<think>[\s\S]*?</think>", "", content).strip()
            citations = response.get("citations", [])

            research_data = {"content": content, "citations": citations}
            meta = {"model": "sonar-deep-research", "command": "perplexity research", "duration_ms": duration_ms}

            if not grok_challenge:
                return self.success(research_data, meta)

            # Chain to Grok challenger
            xai_cmd = XaiSearchCommand()
            challenge_result = xai_cmd.execute(query=query, results_to_challenge=content)

            result = {"research": research_data}
            if challenge_result.get("status") == "success":
                result["challenge"] = challenge_result["data"]
                meta["challenge_duration_ms"] = challenge_result.get("meta", {}).get("duration_ms")
            else:
                result["challenge"] = {"error": challenge_result.get("error", "Unknown error")}

            return self.success(result, meta)

        except requests.exceptions.Timeout:
            return self.error("timeout", f"Request timed out after {self.timeout_ms}ms")
        except requests.exceptions.RequestException as e:
            return self.error("api_error", str(e))
        except (KeyError, IndexError) as e:
            return self.error("parse_error", f"Unexpected response format: {e}")

    def _call_api(self, api_key: str, query: str) -> dict:
        response = requests.post(
            "https://api.perplexity.ai/chat/completions",
            headers={
                "Authorization": f"Bearer {api_key}",
                "Content-Type": "application/json",
            },
            json={
                "model": "sonar-deep-research",
                "messages": [{"role": "user", "content": query}],
            },
            timeout=self.timeout_ms / 1000,
        )
        response.raise_for_status()
        return response.json()


# Auto-discovered by envoy.py
COMMANDS = {
    "research": PerplexityResearchCommand,
}

File: .allhands/harness/src/commands/perplexity.ts
/**
 * Perplexity Commands (Agent-Facing)
 *
 * Web search with citations using Perplexity's sonar-pro model.
 * Matches the configuration used by the Perplexity MCP server.
 *
 * Commands:
 * - ah perplexity research <query> - Web search with citations
 *   --challenge: Challenge findings using Grok X/Twitter search
 */

import { Command } from 'commander';
import { tracedAction } from '../lib/base-command.js';

interface PerplexityResponse {
  choices?: Array<{
    message?: {
      content?: string;
    };
  }>;
  citations?: string[];
}

interface GrokResponse {
  choices?: Array<{
    message?: {
      content?: string;
    };
  }>;
  citations?: string[];
}

const GROK_CHALLENGER_PROMPT = `You are a critical research challenger. Given research findings, search X to:

1. CHALLENGE: Find contradicting opinions, failed implementations, known issues
2. ALTERNATIVES: Surface newer/better tools the research may have missed
3. TRENDS: Identify emerging patterns that could affect the recommendations
4. SENTIMENT: Gauge real developer satisfaction vs marketing claims
5. DISCUSSIONS: Find where the best practitioners are discussing this topic

Be skeptical. Surface what the research missed or got wrong. Focus on recent posts (last 6 months).`;

const GROK_TIMEOUT = 120000;

const PERPLEXITY_TIMEOUT = parseInt(process.env.PERPLEXITY_TIMEOUT_MS ?? '60000', 10);

export function register(program: Command): void {
  const perplexity = program
    .command('perplexity')
    .description('Web search with citations');

  // ah perplexity research
  perplexity
    .command('research <query>')
    .description('Web search with citations (sonar-pro model)')
    .option('--json', 'Output as JSON')
    .action(tracedAction('perplexity research', async (query: string, options: { json?: boolean }) => {
      const apiKey = process.env.PERPLEXITY_API_KEY;
      if (!apiKey) {
        if (options.json) {
          console.log(JSON.stringify({ success: false, error: 'PERPLEXITY_API_KEY not set' }));
        } else {
          console.error('Error: PERPLEXITY_API_KEY not set in environment');
        }
        process.exit(1);
      }

      try {
        const response = await callPerplexityApi(apiKey, query);

        const content = response.choices?.[0]?.message?.content ?? '';
        const citations = response.citations ?? [];

        if (options.json) {
          console.log(JSON.stringify({
            success: true,
            query,
            content,
            citations,
          }, null, 2));
          return;
        }

        console.log('Research Results:');
        console.log();
        console.log(content);

        if (citations.length > 0) {
          console.log();
          console.log('Citations:');
          for (let i = 0; i < citations.length; i++) {
            console.log(`  [${i + 1}] ${citations[i]}`);
          }
        }
      } catch (e) {
        const error = e instanceof Error ? e.message : String(e);
        if (options.json) {
          console.log(JSON.stringify({ success: false, error }));
        } else {
          console.error(`Error: ${error}`);
        }
        process.exit(1);
      }
    }));
}

async function callPerplexityApi(apiKey: string, query: string): Promise<PerplexityResponse> {
  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), PERPLEXITY_TIMEOUT);

  try {
    const response = await fetch('https://api.perplexity.ai/chat/completions', {
      method: 'POST',
      headers: {
        Authorization: `Bearer ${apiKey}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        model: 'sonar-pro',
        messages: [{ role: 'user', content: query }],
      }),
      signal: controller.signal,
    });

    if (!response.ok) {
      throw new Error(`HTTP ${response.status}: ${await response.text()}`);
    }

    return (await response.json()) as PerplexityResponse;
  } finally {
    clearTimeout(timeout);
  }
}
